File: web/src/app/__tests__/contract.test.ts
// 前后端载荷契约测试 (T5.3)
//
// 目的：
// 校验 web/src/app/types.ts 声明的前端接口与后端 Go HTTP/SSE 序列化产物 (contract_fixtures.json) 的强一致性。
// 任何字段大小写拼写、camelCase 命名、枚举值定义不一致均会在此报错，阻止前后端静默漂移。

import { describe, expect, it } from "vitest";
import fixtures from "./contract_fixtures.json";
import type { TurnEventType } from "../api";
import type {
  AcceptResult,
  CheckResult,
  DeriveResult,
  ItemInstance,
  MemoryUsage,
  MemoryView,
  Option,
  PromiseItem,
  RelationValue,
  SessionView,
  TextBlock,
  TurnContent,
} from "../types";

describe("前后端载荷契约测试 (Contract Verification)", () => {
  it("协议版本号对齐", () => {
    expect(fixtures.version).toBe(1);
  });

  it("SessionView 核心字段及结构对齐", () => {
    const sv = fixtures.sessionView as unknown as SessionView;
    expect(sv.sessionId).toBe("sess_contract_1");
    expect(sv.characterId).toBe("char_1");
    expect(sv.title).toBe("风暴前夕");
    expect(sv.rootNodeId).toBe("node_root");
    expect(sv.branch.branchId).toBe("branch_main");
    expect(sv.branch.headNodeId).toBe("node_turn_1");
    expect(sv.branch.version).toBe(1);

    expect(sv.headNode.nodeId).toBe("node_turn_1");
    expect(sv.headNode.turnNumber).toBe(1);
    expect(sv.headNode.kind).toBe("turn");

    expect(sv.secrets).toBeDefined();
    expect(sv.secrets![0].secretId).toBe("sec_1");
    expect(sv.secrets![0].revealed).toBe(true);

    expect(sv.outline).toBeDefined();
    expect(sv.outline!.milestones[0].milestoneId).toBe("m_1");
    expect(sv.outline!.goals[0].goalId).toBe("g_1");
    expect(sv.outline!.scene).toBe("古堡大门");

    expect(sv.director).toBeDefined();
    expect(sv.director!.status).toBe("active");
    expect(sv.director!.currentBeatId).toBe("beat_1");

    expect(sv.actions).toBeDefined();
    expect(sv.actions![0].actionId).toBe("act_search");
    expect(sv.actions![0].dc).toBe(12);

    expect(sv.activeSummary).toBeDefined();
    expect(sv.activeSummary!.summaryId).toBe("sum_1");
  });

  it("WorldState 投影状态字段与命名对齐", () => {
    const sv = fixtures.sessionView as unknown as SessionView;
    const state = sv.state as any;

    expect(state.relationships).toBeDefined();
    const rel: RelationValue = state.relationships["npc_1"];
    expect(rel.affection).toBe(10);
    expect(rel.trust).toBe(15);
    expect(rel.alertness).toBe(0);

    expect(state.items).toBeDefined();
    const item: ItemInstance = state.items["item_1"];
    expect(item.instanceId).toBe("item_1");
    expect(item.keepsake).toBe(true);

    expect(state.promises).toBeDefined();
    const prom: PromiseItem = state.promises["prom_1"];
    expect(prom.promiseId).toBe("prom_1");
    expect(["proposed", "active", "fulfilled", "broken", "cancelled"]).toContain(prom.state);
  });

  it("TurnContent 正文块与枚举对齐", () => {
    const tc = fixtures.turnContent as unknown as TurnContent;
    expect(tc.inputKind).toBe("dialogue");
    expect(tc.inputText).toBe("请问前面有什么危险？");

    expect(tc.blocks.length).toBe(3);
    const validKinds: TextBlock["kind"][] = ["narration", "dialogue", "inner_monologue"];
    for (const b of tc.blocks) {
      expect(validKinds).toContain(b.kind);
    }
    expect(tc.blocks[1].speakerId).toBe("char_1");

    expect(tc.options.length).toBe(2);
    const validIntents: Option["intent"][] = ["aggressive", "clever", "emotional", "chaotic"];
    for (const opt of tc.options) {
      expect(validIntents).toContain(opt.intent);
    }

    expect(tc.checks).toBeDefined();
    const chk = tc.checks![0] as CheckResult;
    expect(chk.rollId).toBe("roll_1");
    expect(chk.actionId).toBe("act_search");
    expect(chk.attributeModifier).toBe(2);
    expect(chk.natural).toBe(15);
    expect(chk.total).toBe(17);
    expect(chk.outcome).toBe("success");

    expect(tc.mood).toBeDefined();
    expect(tc.mood!.characterId).toBe("char_1");
    expect(tc.mood!.moodCode).toBe("vigilant");
  });

  it("MemoryView 结构与时间轴字段对齐", () => {
    const mem = fixtures.memoryView as unknown as MemoryView;
    expect(mem.memoryId).toBe("mem_contract_1");
    expect(mem.sourceNodeId).toBe("node_turn_1");
    expect(["observed", "reported", "inferred", "secret"]).toContain(mem.kind);
    expect(mem.subjectKey).toBe("char_1.badge");
    expect(mem.createdTurn).toBe(1);
    expect(mem.validFromTurn).toBe(1);
    expect(mem.effective).toBe(true);

    const usage = fixtures.memoryUsage as unknown as MemoryUsage;
    expect(usage.used).toBe(1);
    expect(usage.limit).toBe(200);
    expect(["Normal", "Notice", "Degraded", "Critical"]).toContain(usage.tier);
  });

  it("AcceptResult 与 DeriveResult 契约字段对齐", () => {
    const acc = fixtures.acceptResult as unknown as AcceptResult;
    expect(acc.turnId).toBe("turn_new_1");
    expect(acc.status).toBe("processing");
    expect(acc.statusUrl).toBe("/api/v1/turns/turn_new_1");
    expect(acc.eventsUrl).toBe("/api/v1/turns/turn_new_1/events");

    const der = fixtures.deriveResult as unknown as DeriveResult;
    expect(der.branchId).toBe("branch_sub_1");
    expect(der.turnId).toBe("turn_new_2");
    expect(der.branch.branchId).toBe("branch_sub_1");
  });

  it("API Error 契约对齐", () => {
    const err = fixtures.apiError as { code: string; message: string; retryable: boolean; turnId?: string };
    expect(err.code).toBe("INVALID_ARGUMENT");
    expect(err.message).toBe("参数校验失败");
    expect(err.retryable).toBe(false);
    expect(err.turnId).toBe("turn_err_1");
  });

  it("SSE 实时流事件命名与载荷形态对齐", () => {
    const sse = fixtures.sseEvents;
    const recognizedTurnEvents: TurnEventType[] = [
      "turn.started",
      "turn.thinking",
      "block.delta",
      "block.appended",
      "turn.truncated",
      "turn.awaiting_continuation",
      "turn.awaiting_approval",
      "turn.committed",
      "turn.cancelled",
      "turn.failed",
      "turn.conflicted",
    ];

    for (const evName of Object.keys(sse)) {
      expect(recognizedTurnEvents).toContain(evName as TurnEventType);
    }

    const delta = sse["block.delta"] as { attemptId: string; seq: number; kind: string; delta: string };
    expect(delta.attemptId).toBe("att_1");
    expect(delta.seq).toBe(1);
    expect(delta.kind).toBe("narration");
    expect(delta.delta).toBe("森林深处的阴影中，");

    const appended = sse["block.appended"] as { frameSeq: number; frame: { v: number; kind: string; text: string } };
    expect(appended.frameSeq).toBe(1);
    expect(appended.frame.v).toBe(1);
    expect(appended.frame.kind).toBe("narration");

    const failed = sse["turn.failed"] as { code: string; message: string };
    expect(failed.code).toBe("LLM_GATEWAY_TIMEOUT");
  });
});

File: internal/adapters/http/contract_test.go
package http_test

import (
	"encoding/json"
	"os"
	"path/filepath"
	"testing"
	"time"

	"tavernagent/internal/application"
	"tavernagent/internal/domain"
	"tavernagent/internal/protocol"
)

// TestExportAndVerifyPayloadContracts (T5.3 前后端载荷契约测试)
//
// 目的：
// 锁定 Go 后端 HTTP 与 SSE 载荷结构，输出到 web/src/app/__tests__/contract_fixtures.json，
// 并由前端单元测试 web/src/app/__tests__/contract.test.ts 读取校验。
// 防止前后端类型在缺乏 protobuf 的情况下产生静默漂移（字段名大写/驼峰不符、枚举值漂移、字段缺失）。
func TestExportAndVerifyPayloadContracts(t *testing.T) {
	fixedTime := time.Date(2026, 9, 13, 10, 0, 0, 0, time.UTC)

	// 1. SessionView 载荷
	sampleState := map[string]any{
		"relationships": map[string]any{
			"npc_1": map[string]any{"affection": 10, "trust": 15, "alertness": 0},
		},
		"items": map[string]any{
			"item_1": map[string]any{
				"instanceId": "item_1",
				"name":       "旧怀表",
				"quantity":   1,
				"keepsake":   true,
			},
		},
		"promises": map[string]any{
			"prom_1": map[string]any{
				"promiseId":    "prom_1",
				"content":      "遵守约定",
				"sourceNodeId": "node_root",
				"state":        "active",
			},
		},
		"characters": map[string]any{
			"char_1": map[string]any{
				"characterId": "char_1",
				"name":        "艾莲娜",
			},
		},
		"moods": map[string]any{
			"char_1": map[string]any{
				"moodCode": "calm",
				"text":     "心如止水",
			},
		},
		"scene": map[string]any{
			"sceneId": "scene_gates",
			"title":   "古堡大门",
		},
	}
	stateJSON, err := json.Marshal(sampleState)
	if err != nil {
		t.Fatal(err)
	}

	sessionView := application.SessionView{
		SessionID:   "sess_contract_1",
		CharacterID: "char_1",
		Title:       "风暴前夕",
		RootNodeID:  "node_root",
		Branch: application.BranchView{
			BranchID:   "branch_main",
			Name:       "主分支",
			HeadNodeID: "node_turn_1",
			Version:    1,
		},
		ViewNodeID: "node_turn_1",
		HeadNode: &domain.PlotNode{
			NodeID:      "node_turn_1",
			SessionID:   "sess_contract_1",
			ParentID:    "node_root",
			Kind:        domain.NodeKindTurn,
			Depth:       1,
			TurnNumber:  1,
			ContentJSON: `{"inputKind":"dialogue","inputText":"你好"}`,
			CreatedAt:   fixedTime,
		},
		Nodes: []*domain.PlotNode{
			{
				NodeID:      "node_root",
				SessionID:   "sess_contract_1",
				Kind:        domain.NodeKindRoot,
				Depth:       0,
				TurnNumber:  0,
				ContentJSON: `{}`,
				CreatedAt:   fixedTime.Add(-10 * time.Minute),
			},
		},
		State: json.RawMessage(stateJSON),
		Secrets: []application.SecretView{
			{
				SecretID: "sec_1",
				Title:    "失落遗迹",
				Order:    1,
				Content:  "神殿藏有远古符文石。",
				Revealed: true,
			},
		},
		Outline: &application.OutlineView{
			Milestones: []application.OutlineEntry{
				{MilestoneID: "m_1", Description: "穿越黑森林"},
			},
			Goals: []application.GoalEntry{
				{GoalID: "g_1", CharacterID: "char_1", Text: "找到失落的石板"},
			},
			Scene: "古堡大门",
		},
		ActiveSummary: &domain.SummaryArtifact{
			SummaryID:  "sum_1",
			FromNodeID: "node_root",
			ToNodeID:   "node_turn_1",
			SourceHash: "hash_abcdef",
			Text:       "主角踏入了大门。",
		},
		Actions: []application.ActionView{
			{ActionID: "act_search", Label: "搜寻线索", Attribute: "perception", DC: 12},
		},
		Director: &domain.DirectorSummary{
			Title:            "第一幕：启程",
			RevisionID:       "rev_1",
			Status:           "active",
			CurrentBeatID:    "beat_1",
			CurrentBeatTitle: "初遇",
			Completed:        0,
			Total:            3,
		},
		HasMore: false,
	}

	// 2. TurnContent 载荷
	turnContent := domain.TurnContent{
		InputKind: "dialogue",
		InputText: "请问前面有什么危险？",
		Checks: []domain.CheckResult{
			{
				RollID:          "roll_1",
				ActionID:        "act_search",
				Attribute:       "perception",
				AttributeMod:    2,
				Natural:         15,
				Total:           17,
				DC:              12,
				Outcome:         domain.OutcomeSuccess,
				RulesetVer:      "v1",
				PermanentEffect: "感知敏锐",
			},
		},
		Blocks: []domain.TextBlock{
			{
				Kind: "narration",
				Text: "远方传来低沉的咆哮。",
			},
			{
				Kind:      "dialogue",
				SpeakerID: "char_1",
				Text:      "小心，前面有魔兽出没。",
			},
			{
				Kind: "inner_monologue",
				Text: "（它比预想的还要敏捷……）",
			},
		},
		Options: []domain.Option{
			{
				OptionID: "opt_1",
				Intent:   "clever",
				Text:     "攀上树枝观察周围地形",
			},
			{
				OptionID: "opt_2",
				Intent:   "aggressive",
				Text:     "拔剑戒备准备正面突击",
			},
		},
		Mood: &domain.Mood{
			CharacterID: "char_1",
			MoodCode:    "vigilant",
			Text:        "警惕环视四周",
		},
	}

	// 3. MemoryView 载荷
	memoryView := application.MemoryView{
		MemoryRecord: &domain.MemoryRecord{
			MemoryID:      "mem_contract_1",
			SourceNodeID:  "node_turn_1",
			Kind:          "observed",
			OwnerIDs:      []string{"char_1"},
			Content:       "艾莲娜佩戴着一枚带有雄鹰印记的银制徽章。",
			EntityIDs:     []string{"char_1"},
			Confidence:    0.95,
			Pinned:        false,
			SubjectKey:    "char_1.badge",
			CreatedTurn:   1,
			ValidFromTurn: 1,
		},
		Effective: true,
		Protected: false,
	}

	memoryUsage := domain.MemoryUsage{
		Used:        1,
		Protected:   0,
		Limit:       200,
		Reclaimable: 0,
		Headroom:    199,
		Tier:        "Normal",
	}

	// 4. AcceptResult & DeriveResult
	acceptResult := map[string]any{
		"turnId":    "turn_new_1",
		"status":    "processing",
		"statusUrl": "/api/v1/turns/turn_new_1",
		"eventsUrl": "/api/v1/turns/turn_new_1/events",
	}

	deriveResult := map[string]any{
		"branch": application.BranchView{
			BranchID:   "branch_sub_1",
			Name:       "新分支",
			HeadNodeID: "node_root",
			Version:    1,
		},
		"branchId":  "branch_sub_1",
		"turnId":    "turn_new_2",
		"status":    "processing",
		"statusUrl": "/api/v1/turns/turn_new_2",
		"eventsUrl": "/api/v1/turns/turn_new_2/events",
	}

	// 5. API Error 载荷 (与 writeError 对齐)
	apiError := map[string]any{
		"code":      "INVALID_ARGUMENT",
		"message":   "参数校验失败",
		"retryable": false,
		"turnId":    "turn_err_1",
	}

	// 6. SSE 实时事件载荷
	sseEvents := map[string]any{
		"turn.thinking": map[string]any{
			"attemptId": "att_1",
			"thinking":  "正在构思剧情转折...",
		},
		"block.delta": map[string]any{
			"attemptId": "att_1",
			"seq":       1,
			"kind":      "narration",
			"speakerId": nil,
			"delta":     "森林深处的阴影中，",
		},
		"block.appended": map[string]any{
			"frameSeq": 1,
			"frame": protocol.BlockFrame{
				V:    1,
				Seq:  1,
				Type: protocol.FrameBlock,
				Kind: protocol.BlockNarration,
				Text: "森林深处的阴影中，隐约有红色的眼睛亮起。",
			},
		},
		"turn.started":   map[string]any{},
		"turn.committed": map[string]any{},
		"turn.cancelled": map[string]any{},
		"turn.failed": map[string]any{
			"code":    "LLM_GATEWAY_TIMEOUT",
			"message": "模型供应商响应超时",
		},
		"turn.conflicted": map[string]any{},
	}

	fixtures := map[string]any{
		"version":      protocol.Version,
		"sessionView":  sessionView,
		"turnContent":  turnContent,
		"memoryView":   memoryView,
		"memoryUsage":  memoryUsage,
		"acceptResult": acceptResult,
		"deriveResult": deriveResult,
		"apiError":     apiError,
		"sseEvents":    sseEvents,
	}

	// 验证可正常序列化
	data, err := json.MarshalIndent(fixtures, "", "  ")
	if err != nil {
		t.Fatalf("marshal fixtures failed: %v", err)
	}

	// 输出到 web/src/app/__tests__/contract_fixtures.json 供前端测试加载
	targetPath := filepath.Join("..", "..", "..", "web", "src", "app", "__tests__", "contract_fixtures.json")
	if err := os.WriteFile(targetPath, data, 0o644); err != nil {
		t.Fatalf("write fixture file %s failed: %v", targetPath, err)
	}
}
